### ISSUE TITLE
How to reduce the randomness in training so that the training results are repeatable?

### ISSUE THREAD

**Issue by whyygug** [nianticlabs/monodepth2#303]
If we train multiple models with exactly the same configuration, they should behave the same, but I run into problems here.
I trained several different models with the same configuration: `python train.py --model_name mono_model` and their performance was basically the same, but there were some differences.
For example, the abs_rel has been fluctuating between 0.113 and 0.116, while the sq_rel has fluctuated even more. There are also more differences in the visualization of depth maps, some models can distinguish small objects, while others can't.
I've added the following code to try to reduce the randomness a bit, but with little success：
```
        #in the beginning of trainer.py
        self.seed = 0
        torch.manual_seed(self.seed)
        torch.cuda.manual_seed(self.seed)
        torch.cuda.manual_seed_all(self.seed)
        np.random.seed(self.seed)
        torch.backends.cudnn.deterministic = True
        torch.backends.cudnn.benchmark = False
```
and
```
        #in the beginning of mono_dataset.py, to make sure the model selects the same data augmentation
        random.seed(0)
```
I trained all of these models to use the same configuration and Settings, but why should they be different?
Can you give me some suggestions to reduce the randomness during training?
Many thanks！

**Comment by mrharicot** [comment]
Hi,
Great question!
I see you have already found the[ pytorch doc on reproducibility](https://pytorch.org/docs/stable/notes/randomness.html)
You seem to be missing a ```torch.set_deterministic(True)``` as well.
Note that this will likely slow down training significantly.

At this point the main source of randomness comes from multithreading in the dataloader: different augmentations will be applied to different samples.
It is unclear to me whether this can be easily solved (see [this issue](https://discuss.pytorch.org/t/is-there-a-way-to-fix-the-random-seed-of-every-workers-in-dataloader/21687/10)).

The main change which I think would improve result stability would be to increase the batch size.
Batchnorm is used in the encoder and it is known to behave unexpectedly on small batches (less than 32) as the variance computation will be fairly noisy.
Using amp (mixed precision) from more recent pytorch releases would allow you to train with ~2X larger batches on recent GPUs.
Another (complementary) option would be to replace the batchnorm with groupnorm which behaves much better with smaller batch sizes.

Let us know if you make progress on that front, reproducibility has been an issue when developing monodepth2 from the begining.

I hope this helps!

**Comment by whyygug** [comment]
Hi, thanks for the reply.
The machine I use for model training is a bit old (Tesla P40), and the CUDA version is 10.0, according to the PyTorch website, the latest version of PyTorch available for this CUDA is 1.2.0，which doesn't seem to support `torch.set_deterministic(True)`. Unfortunately, I can't update the CUDA version because it might affect other people who are using the machine. So I only use the code above.  These operations do significantly slow down the training speed，(monocular training now takes around 15 hours instead of 12 hours)

The following content is built on adding the code from [pytorch doc on reproducibility 1.2.0](https://pytorch.org/docs/1.2.0/notes/randomness.html?highlight=reproducibility)

As to the dataloader, I noticed that when the structure of the entire model is determined, the order of dataloader's output do not change, because when I looked at the results of the training set in Tensorboard, the images that were shown and the order in which they were shown remained the same, but when I change any part of the model (such as the network structure, or the loss function) the images that are shown will become different, and then they will be fixed again until I modify the model again. So even if the random seed of the dataloader is fixed, the order in which the dataloader output images will still change, maybe this phenomenon is related to time?

Group normalization is really appealing, but it will result in not being able to use pre-trained weights of ResNet. （I couldn't find a  pre-trained ResNet that use group normalization for PyTorch）

Mixed precision is also a good idea, but only new RTX GPUs support this operation, perhaps I'll try it later when I move on to a better device.

**Comment by whyygug** [comment]
@mrharicot 
I add group norm in decoder layers, but this seems to make the training more unstable. 
I change the original code from：
https://github.com/nianticlabs/monodepth2/blob/ab2a1bf7d45ae53b1ae5a858f4451199f9c213b3/layers.py#L106-L118

to：

```
class ConvBlock(nn.Module):

    def __init__(self, in_channels, out_channels):
        super(ConvBlock, self).__init__()

        self.conv = Conv3x3(in_channels, out_channels)
        self.nonlin = nn.ELU(inplace=True)
        self.normalize = torch.nn.GroupNorm(16, out_channels)

    def forward(self, x):
        out = self.conv(x)
        out = self.normalize(out)
        out = self.nonlin(out)
        return out
```

then， Tensorboard for training set shows this, seems more unstable.
![image](https://user-images.githubusercontent.com/31730585/105172288-5b302c00-5b5a-11eb-9223-28e7a8771a06.png)


Tensorboard for validation set shows this：
![image](https://user-images.githubusercontent.com/31730585/105172430-9599c900-5b5a-11eb-939f-173587a57a49.png)


The evaluation on the test set is a little worse than before.
This is not quite as expected, because GN is supposed to accelerate the convergence of the model and lead to better results.

**Comment by mrharicot** [comment]
Hi,
I have always struggled to get good stability with normalization in the decoder, you could try using InstanceNorm2D.
Moreover I would probably switch back to ReLU or LeakyReLU if using normalization.
Finally the usual order is Conv/BN/ReLU, I believe you swapped BN and Relu.

I hope this helps!

**Comment by SBaokun** [comment]
@whyygug hello，i met the same problem, do you hava new solutation? thanks!
